theano/compile/builders.py

```diff
@@ -1,44 +1,121 @@
+"""Define new Ops from existing Ops"""
 from __future__ import absolute_import, print_function, division
+from functools import reduce, partial
+from collections import OrderedDict
+
 import theano
 from theano import gof
 from theano.compat import izip
 from theano.compile.function_module import orig_function
-from theano.compile import SharedVariable, rebuild_collect_shared
+from theano.compile import SharedVariable, rebuild_collect_shared, optdb
 from theano.gof import ops_with_inner_function
 from theano.gof.graph import io_connection_pattern
-
-from functools import reduce
+from theano.gof.null_type import NullType
+from theano.gradient import DisconnectedType
 
 
 class OpFromGraph(gof.Op):
     """
-    This creates an `Op` from inputs and outputs lists of variables.
-
-    The signature is similar to theano.function() and the resulting
-    `Op`'s perform will do the same operation as::
+    This creates an ``Op`` from inputs and outputs lists of variables.
+    The signature is similar to :func:`theano.function <theano.function>`
+    and the resulting ``Op``'s perform will do the same operation as::
 
         orig_function(inputs, outputs, **kwargs)
 
-    TODO:
+    Currently does not support ``updates`` or ``givens`` argument.
+
+    Parameters
+    ----------
+
+    inputs: list of :class:`Variable <theano.gof.Variable>`
+
+    outputs: list of :class:`Variable <theano.gof.Variable>`
+
+    inline: bool, optional
+        Defaults to ``False``
+
+        ``True`` : Cause the Op's original graph being used during
+        compilation, the Op will not be visible in the compiled
+        graph but rather its internal graph.
+
+        ``False`` : will use a pre-compiled function inside.
+
+    grad_overrides : single or list of {0, None, Ellipsis, OpFromGraph, callable}, optional
+        Defaults to ``None``.
+
+        ``None`` : No value, gives NullType()
+        ``0`` : zero value, gives DisconnectedType()
+        ``...`` : Do not override, use default grad() result
+
+        OpFromGraph instance : Override with another OpFromGraph, should
+        accept inputs as the same order and types of "inputs" and "output_grads"
+        arguments as one would specify in grad() method.
+
+        callable : similar to OpFromGraph instance, must return list of
+        :class:`Variable <theano.gof.Variable>`.
+
+        list: Each OpFromGraph/callable must return a single
+        :class:`Variable <theano.gof.Variable>`. Each list element corresponds to gradient of
+        a specific input, length of list must be equal to number of inputs.
+
+    rop_overrides : single or list of {0, None, Ellipsis, OpFromGraph, callable}, optional
+        Defaults to ``None``.
+
+        ``None`` : No value, gives NullType()
+        ``0`` : zero value, gives zeros_like(...)
+        ``...`` : Do not override, use default R_op() result
```

Comment:
For now, I think string `"default"` would suffice. In future, I think there should be global singletons like `theano.Default`, `theano.Automatic` (I lend the idea from Mathematica), and then make all theano interfaces uniform. Just a personal idea.